python/ray/_raylet.pyx

```diff
@@ -357,56 +276,41 @@ cdef c_vector[c_string] string_vector_from_list(list string_list):
         out.push_back(s)
     return out
 
-
-cdef:
-    c_string pickle_metadata_str = PICKLE_BUFFER_METADATA
-    shared_ptr[CBuffer] pickle_metadata = dynamic_pointer_cast[
-        CBuffer, LocalMemoryBuffer](
-        make_shared[LocalMemoryBuffer](
-            <uint8_t*>(pickle_metadata_str.data()),
-            pickle_metadata_str.size(), True))
-    c_string raw_meta_str = RAW_BUFFER_METADATA
-    shared_ptr[CBuffer] raw_metadata = dynamic_pointer_cast[
-        CBuffer, LocalMemoryBuffer](
-        make_shared[LocalMemoryBuffer](
-            <uint8_t*>(raw_meta_str.data()),
-            raw_meta_str.size(), True))
-
-cdef void prepare_args(list args, c_vector[CTaskArg] *args_vector):
+cdef void prepare_args(
+        CoreWorker core_worker, list args, c_vector[CTaskArg] *args_vector):
     cdef:
-        c_string pickled_str
-        const unsigned char[:] buffer
         size_t size
+        int64_t put_threshold
         shared_ptr[CBuffer] arg_data
-        shared_ptr[CBuffer] arg_metadata
 
-    # TODO be consistent with store_task_outputs
+    worker = ray.worker.global_worker
+    put_threshold = RayConfig.instance().max_direct_call_object_size()
     for arg in args:
         if isinstance(arg, ObjectID):
             args_vector.push_back(
                 CTaskArg.PassByReference((<ObjectID>arg).native()))
-        elif not ray._raylet.check_simple_value(arg):
-            args_vector.push_back(
-                CTaskArg.PassByReference((<ObjectID>ray.put(arg)).native()))
-        elif type(arg) is bytes:
-            buffer = arg
-            size = buffer.nbytes
-            arg_data = dynamic_pointer_cast[CBuffer, LocalMemoryBuffer](
-                make_shared[LocalMemoryBuffer](
-                    <uint8_t*>(&buffer[0]), size, True))
-            args_vector.push_back(
-                CTaskArg.PassByValue(
-                    make_shared[CRayObject](arg_data, raw_metadata)))
+
         else:
-            buffer = pickle.dumps(
-                arg, protocol=pickle.HIGHEST_PROTOCOL)
-            size = buffer.nbytes
-            arg_data = dynamic_pointer_cast[CBuffer, LocalMemoryBuffer](
-                    make_shared[LocalMemoryBuffer](
-                        <uint8_t*>(&buffer[0]), size, True))
-            args_vector.push_back(
-                CTaskArg.PassByValue(
-                    make_shared[CRayObject](arg_data, pickle_metadata)))
+            serialized_arg = worker.get_serialization_context().serialize(arg)
+            size = serialized_arg.total_bytes
+
+            # TODO(edoakes): any objects containing ObjectIDs are spilled to
```

Comment:
This could really hurt performance for some applications. One option is to instead just put these in the local memory store rather than plasma and then go through the normal refcounting logic

Reply:
Yeah, there is a TODO about this [here](https://github.com/ray-project/ray/blob/master/src/ray/core_worker/core_worker.cc#L1095).

Not too worried about it, though, because it should at least be as fast as before, when all objects went through plasma.

Reply:
Actually I don't think that's the case - previously we were inlining "simple" objects by using that very complicated check and then pickling them with vanilla `pickle`